### PR TITLE
[SPARK-8978][Streaming] Implements the DirectKafkaRateController

### PR DESCRIPTION
**Pull request by huitseeker**
This depends on #16

### CODE REVIEW

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala [hunk 1]
```diff
@@ -73,12 +73,14 @@ class DirectKafkaInputDStream[
...
   protected val kc = new KafkaCluster(kafkaParams)
 
-  protected val maxMessagesPerPartition: Option[Long] = {
-    val ratePerSec = context.sparkContext.getConf.getInt(
+  private val ratePerSec: Int = context.sparkContext.getConf.getInt(
       "spark.streaming.kafka.maxRatePerPartition", 0)
-    if (ratePerSec > 0) {
+  protected def maxMessagesPerPartition: Option[Long] = {
+    val estimatedRate = rateController.getLatestRate().toInt
+    val effectiveRatePerSec = if (estimatedRate > 0) (ratePerSec min estimatedRate) else ratePerSec
```
Comment (dragos) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala]: The estimated rate is per stream, not per Kafka partition (like the configuration parameter above). Shouldn't this be accounted for in here?

**Comment by typesafe-tools** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://ci.typesafe.com/job/ghprb-spark-multi-conf/48/

**Build Log**
_last 10 lines_

```
[...truncated 11 lines...]
 > git rev-parse refs/remotes/origin/origin/pr/18/merge^{commit} # timeout=10
Checking out Revision 0f753bd12042de2eab2dd3ac41b934716c795d17 (refs/remotes/origin/pr/18/merge)
 > git config core.sparsecheckout # timeout=10
 > git checkout -f 0f753bd12042de2eab2dd3ac41b934716c795d17
First time build. Skipping changelog.
Triggering ghprb-spark-multi-conf » Spark-Ora-JDK7-PV,2.10
ghprb-spark-multi-conf » Spark-Ora-JDK7-PV,2.10 completed with result FAILURE
Notifying upstream projects of job completion
Setting status of 63593b2e1c18eab652c677bfa005a120bfbce019 to FAILURE with url http://ci.typesafe.com/job/ghprb-spark-multi-conf/48/ and message: Merged build finished.
```

Test FAILed.

**Comment by typesafe-tools** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://ci.typesafe.com/job/ghprb-spark-multi-conf/49/

**Build Log**
_last 10 lines_

```
[...truncated 11 lines...]
 > git rev-parse refs/remotes/origin/origin/pr/18/merge^{commit} # timeout=10
Checking out Revision 0f753bd12042de2eab2dd3ac41b934716c795d17 (refs/remotes/origin/pr/18/merge)
 > git config core.sparsecheckout # timeout=10
 > git checkout -f 0f753bd12042de2eab2dd3ac41b934716c795d17
First time build. Skipping changelog.
Triggering ghprb-spark-multi-conf » Spark-Ora-JDK7-PV,2.10
ghprb-spark-multi-conf » Spark-Ora-JDK7-PV,2.10 completed with result FAILURE
Notifying upstream projects of job completion
Setting status of bb436a2381af2bc4033c90bb1697bc2aac3b1e44 to FAILURE with url http://ci.typesafe.com/job/ghprb-spark-multi-conf/49/ and message: Merged build finished.
```

Test FAILed.

**Comment by typesafe-tools** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://ci.typesafe.com/job/ghprb-spark-multi-conf/50/

**Build Log**
_last 10 lines_

```
[...truncated 11 lines...]
 > git rev-parse refs/remotes/origin/origin/pr/18/merge^{commit} # timeout=10
Checking out Revision cd968115e85ebc13dfc6d4ff8c3dbc8622b464cf (refs/remotes/origin/pr/18/merge)
 > git config core.sparsecheckout # timeout=10
 > git checkout -f cd968115e85ebc13dfc6d4ff8c3dbc8622b464cf
First time build. Skipping changelog.
Triggering ghprb-spark-multi-conf » Spark-Ora-JDK7-PV,2.10
ghprb-spark-multi-conf » Spark-Ora-JDK7-PV,2.10 completed with result FAILURE
Notifying upstream projects of job completion
Setting status of 73399ced62f6506de93e8f2ff53c6566b6d14ca3 to FAILURE with url http://ci.typesafe.com/job/ghprb-spark-multi-conf/50/ and message: Merged build finished.
```

Test FAILed.

**Comment by typesafe-tools** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://ci.typesafe.com/job/ghprb-spark-multi-conf/53/

**Build Log**
_last 10 lines_

```
[...truncated 11 lines...]
 > git rev-parse refs/remotes/origin/origin/pr/18/merge^{commit} # timeout=10
Checking out Revision cd968115e85ebc13dfc6d4ff8c3dbc8622b464cf (refs/remotes/origin/pr/18/merge)
 > git config core.sparsecheckout # timeout=10
 > git checkout -f cd968115e85ebc13dfc6d4ff8c3dbc8622b464cf
First time build. Skipping changelog.
Triggering ghprb-spark-multi-conf » Spark-Ora-JDK7-PV,2.10
ghprb-spark-multi-conf » Spark-Ora-JDK7-PV,2.10 completed with result FAILURE
Notifying upstream projects of job completion
Setting status of 0d4710bb9810dff1023487cd2804c2516ea126a8 to FAILURE with url http://ci.typesafe.com/job/ghprb-spark-multi-conf/53/ and message: Merged build finished.
```

Test FAILed.

**Comment by nraychaudhuri** [comment]
ok to test

**Comment by typesafe-tools** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://ci.typesafe.com/job/ghprb-spark-multi-conf/56/

**Build Log**
_last 10 lines_

```
[...truncated 18 lines...]
 > git config core.sparsecheckout # timeout=10
 > git checkout -f 4939a0c1b9ab1d79e36a73f5c98b2ce16aac9b47
First time build. Skipping changelog.
Triggering ghprb-spark-multi-conf » Spark-Ora-JDK7-PV,2.10
Configuration ghprb-spark-multi-conf » Spark-Ora-JDK7-PV,2.10 is still in the queue: Waiting for next available executor on Spark JDK-7 PV (i-a6d3f874)
Configuration ghprb-spark-multi-conf » Spark-Ora-JDK7-PV,2.10 is still in the queue: Waiting for next available executor on Spark-Ora-JDK7-PV
ghprb-spark-multi-conf » Spark-Ora-JDK7-PV,2.10 completed with result FAILURE
Notifying upstream projects of job completion
Setting status of 4318a97b5d56988c9ef9320376fde0f4e8e0da1c to FAILURE with url http://ci.typesafe.com/job/ghprb-spark-multi-conf/56/ and message: Merged build finished.
```

Test FAILed.

**Comment by nraychaudhuri** [comment]
ok to test

**Comment by typesafe-tools** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://ci.typesafe.com/job/ghprb-spark-multi-conf/57/

**Build Log**
_last 10 lines_

```
[...truncated 16 lines...]
 > git rev-parse refs/remotes/origin/origin/pr/18/merge^{commit} # timeout=10
Checking out Revision 4939a0c1b9ab1d79e36a73f5c98b2ce16aac9b47 (refs/remotes/origin/pr/18/merge)
 > git config core.sparsecheckout # timeout=10
 > git checkout -f 4939a0c1b9ab1d79e36a73f5c98b2ce16aac9b47
First time build. Skipping changelog.
Triggering ghprb-spark-multi-conf » Spark-Ora-JDK7-PV,2.10
ghprb-spark-multi-conf » Spark-Ora-JDK7-PV,2.10 completed with result FAILURE
Notifying upstream projects of job completion
Setting status of 4318a97b5d56988c9ef9320376fde0f4e8e0da1c to FAILURE with url http://ci.typesafe.com/job/ghprb-spark-multi-conf/57/ and message: Merged build finished.
```

Test FAILed.

**Pull request closed without merging**